### PR TITLE
[IV-21-22] Objetivo 4

### PR DESCRIPTION
**Pull request by arguellesm**
### Description

This PR:
- Starts implementing business logic via **blacklists** and a **Naive Bayes classifier algorithm**.
- Adds **tests** managed with Pytest for the class `Article`.

### Checklist

* [x]  ¿Hay un camino claro que permita ir desde el código hasta la historia de
   usuario que desarrolla, pasando por el mensaje de commit y el issue
   correspondiente?
* [x]  ¿Se ha intentado cubrir una parte de la lógica de negocio que se
   esté desarrollando?
   _Yes, articles have now a `predict()` method that returns the probability
of that article of being misleading or fake._
* [x]  ¿Hay un producto mínimamente viable que describa lo que se va a entregar en
   este objetivo?
   _[Milestone 0.2.0](https://github.com/arguellesm/faux/milestone/3)._
* [x]  ¿En este producto mínimamente viable se han priorizado unas clases,
   módulos o paquetes más fundamentales frente a otros, movidos a otro
   PMV/Hito?
* [x]  ¿Se han documentado las elecciones de biblioteca de aserciones y de
   *framework* de test?
* [x] ¿Esa documentación incluye criterios de aceptación y también criterios de
búsqueda para las diferentes opciones?
* [x]  ¿Se han seguido los principios F.I.R.S.T. en la creación del test y
   se ha documentado cómo se ha hecho?
* [x]  ¿Se describe claramente en el PMV los tests que hay que pasar para que se
   considere viable?

### CODE REVIEW

**Comment by arguellesm** [comment]
@JJ ready! 👀

File: data/generate-df.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+	with open(path) as f:
+		file = json.load(f)
+
+	if 'author' in file.keys():
+		author = file['author']
+
+	if 'authors' in file.keys():
+		author = file['authors']
+
+	return file['text'], author
```
Comment (JJ) [inline data/generate-df.py:20]: You should really use this in an OO way. You're simply returning an array. Clean code calls for using data structures for its purpose, and arrays were not made for this.

File: docs/dev_tools.md [hunk 2]
```diff
@@ -22,9 +22,10 @@ The main alternatives to Poetry are Pip and Pipenv:
   [PEP 518](https://www.python.org/dev/peps/pep-0518/#file-format)).
 
 
+
 ### [Invoke](https://www.pyinvoke.org/)
 
-Invoke is our task runner of choice. It inherits a minimal yet 
+Invoke is our **task runner** of choice. It inherits a minimal yet
```
Comment (JJ) [inline docs/dev_tools.md:28]: Again, don't tell me your choice. Show me your best effort to choose it, and the criteria you have followed.

File: docs/dev_tools.md [hunk 3]
```diff
@@ -42,3 +43,39 @@ Some alternatives are:
   _doit_ has quite a similar approach to Invoke and uses Python as
```
Comment (JJ) [inline docs/dev_tools.md:43]: ... which is why it's used in Python.

File: docs/dev_tools.md [hunk 4]
```diff
@@ -42,3 +43,39 @@ Some alternatives are:
   _doit_ has quite a similar approach to Invoke and uses Python as 
   well. However, Invoke's community is much bigger, which usually
   translates to more resources and easier troubleshooting.
+
+
+
+ ### [Pytest](https://docs.pytest.org/en/latest/)
+ 
+ Pytest is both our **test framework** and **assertion library** of
```
Comment (JJ) [inline docs/dev_tools.md:51]: No, it's not. It barely has any assertions.

File: docs/dev_tools.md [hunk 5]
```diff
@@ -42,3 +43,39 @@ Some alternatives are:
...
+ choice. These were the criteria used in the decision:
+ 
+ - Well documented, preferably with a wide community behind it.
+ - Already included in Python.
+ - Ease of use.
+ - Simplicity.
+
+When searching for test frameworks and assertion libraries, we searched
+for `python + [term]` and `python + [term] + stackoverflow`. These were
```
Comment (JJ) [inline docs/dev_tools.md:60]: ??? stackoverflow why?

File: docs/dev_tools.md [hunk 6]
```diff
@@ -42,3 +43,39 @@ Some alternatives are:
...
+the main resuls for each category:
+
+- Test frameworks
+  - [Pytest](https://docs.pytest.org/en/latest/)
+  - [Unittest](https://docs.python.org/3/library/unittest.html)
+- Assertion libraries
+  - [Assertpy](https://pypi.org/project/assertpy/)
+  - [Unittest](https://docs.python.org/3/library/unittest.html)
```
Comment (JJ) [inline docs/dev_tools.md:68]: Where have I seen this before? Ah, above. Where it does not belong. You can call it a test framework, but it's not by any means in the same class as pytest.

File: test/article_test.py [hunk 7]
```diff
@@ -0,0 +1,49 @@
+import pytest
+from src.article import Article
+
+
+# test article
+article = Article(headline='A simple headline', content='And a very short content')
+
+
+def test_can_create_object_only_headline():
+    assert Article(headline='A simple headline')
```
Comment (JJ) [inline test/article_test.py:10]: This assertion does not belong to pytest. That means that you're not using pytest as an assertion library as you claim.

**Review by JJ — CHANGES_REQUESTED** [review]
Besides all that's been said, there's no business logic tested here.

File: data/generate-df.py (hunk 1 above)
Comment (arguellesm) [reply data/generate-df.py]: I've modified it to make it cleaner, although this script is not part of the application it self, it's just a tool in case someone wants to replicate how the dataframe was created, as the version from FakeHealth has a different format.

File: docs/dev_tools.md (hunk 2 above)
Comment (arguellesm) [reply docs/dev_tools.md]: The reason for choosing Invoke was that its community seems quite bigger. I find this useful for understanding and troubleshooting issues I might have in the future.

File: docs/dev_tools.md (hunk 6 above)
Comment (arguellesm) [reply docs/dev_tools.md]: I've modified this to include Unittest and the `assert` statement as the assertion library/tools 👍🏼

**Comment by arguellesm** [comment]
> Besides all that's been said, there's no business logic tested here.

I don't understand why. Business logic here includes predicting whether an article might be fake or not and to what extent, and I understand that's tested with:

```python
test_predict_returns_three_values()
test_predict_probability_between_0_and_1()
test_author_score_is_0()
test_source_score_is_0()
```

I guess I'm missing something but I don't know what (?)

Other than that, ready for review @JJ

File: docs/dev_tools.md [hunk 8]
```diff
@@ -2,30 +2,31 @@
...
   which is the official Python file for dependency management (see
   [PEP 518](https://www.python.org/dev/peps/pep-0518/#file-format)).
 
 
+
 ### [Invoke](https://www.pyinvoke.org/)
 
-Invoke is our task runner of choice. It inherits a minimal yet 
-powerful philosophy from GNU Make and is broadly used among the 
+Invoke is our **task runner** of choice. It inherits a minimal yet
```
Comment (JJ) [inline docs/dev_tools.md:28]: You should state your criteria before the tool you've chosen. If you don't, it looks like you're simply justifying what you had decided in advance. Also, not this objective, but the previous one. I should have told you so there.

File: docs/dev_tools.md [hunk 9]
```diff
@@ -34,11 +35,57 @@ Some alternatives are:
...
 
-  _doit_ has quite a similar approach to Invoke and uses Python as 
+  _doit_ has quite a similar approach to Invoke and uses Python as
   well. However, Invoke's community is much bigger, which usually
   translates to more resources and easier troubleshooting.
+
+
+ ### [Pytest](https://docs.pytest.org/en/latest/)
+
+ Pytest is both our **test framework** of
```
Comment (JJ) [inline docs/dev_tools.md:50]: both what?

File: docs/dev_tools.md [hunk 10]
```diff
@@ -34,11 +35,57 @@ Some alternatives are:
...
+ choice. These were the criteria used in the decision:
+
+ - Well documented, preferably with a wide community behind it.
```
Comment (JJ) [inline docs/dev_tools.md:53]: Again, criteria should be announced in advance, and should not be subjective, like "well documented". Are the others badly documented?

File: docs/dev_tools.md [hunk 11]
```diff
@@ -34,11 +35,57 @@ Some alternatives are:
...
+ - Already included in Python.
```
Comment (JJ) [inline docs/dev_tools.md:54]: It's *not* included in Python.

File: docs/dev_tools.md [hunk 12]
```diff
@@ -34,11 +35,57 @@ Some alternatives are:
...
+ - Ease of use.
+ - Simplicity.
+
+When searching for test frameworks, we searched
+for `python + [term]` and `python + [term] + stackoverflow`. These were
+the main resuls:
+
+- Test frameworks
+  - [Pytest](https://docs.pytest.org/en/latest/)
+  - [Unittest](https://docs.python.org/3/library/unittest.html)
```
Comment (JJ) [inline docs/dev_tools.md:64]: This is not a test runner. Pytest includes a test runner. I explained in class what's the concept of test runner, including the fact that it's able to parse TAP output.

File: docs/dev_tools.md [hunk 13]
```diff
@@ -34,11 +35,57 @@ Some alternatives are:
...
+  - [The `assert` statement](https://docs.python.org/3/reference/simple_stmts.html#assert) (which is not really a library)
+
+We first discarded using Assertpy to use as few external libraries
+as possible. Unittest offers many tools although is not the simplest,
+as tests should be part of classes that inherit from `unittest.TestCase`.
+In any case, failed assertions from Unittest provided added information
+on why the condition failed, which will be useful in the future.
+
+To allow simplicity when needed, we'll be also using the `assert`
```
Comment (JJ) [inline docs/dev_tools.md:90]: How do you test exceptions then? Also, why don't you include possible exceptions in your code?

File: pyproject.toml [hunk 14]
```diff
@@ -9,6 +9,8 @@ license = "GPL-3.0"
 python = "^3.9"
 invoke = "^1.6.0"
 pylint = "^2.11.1"
+scikit-learn = "^1.0.1"
```
Comment (JJ) [inline pyproject.toml:12]: Is this used?

File: pyproject.toml [hunk 15]
```diff
@@ -9,6 +9,8 @@ license = "GPL-3.0"
...
+pytest = "^6.2.5"
```
Comment (JJ) [inline pyproject.toml:13]: Is this a production or a dev dependency?

File: src/article.py [hunk 16]
```diff
@@ -89,4 +92,29 @@ def id(self):
...
+        if self.author and self.author in blacklists.UNTRUSTED_AUTHORS:
+            author_score = 1
+
+        # check source
+        if self.source and self.source in backlists.UNTRUSTED_SOURCES:
+            source_score = 1
+
+        # check content
+        file = open(os.path.dirname(__file__)+'/../model/mnb.pkl', 'rb')
+        model = pickle.load(file)
```
Comment (JJ) [inline src/article.py:115]: This is not your own logic, it's a model you've picked up from elsewhere. You're not really testing your logic, but some other logic. This objective is about whitelist testing, and about testing your own business logic.

File: src/article.py [hunk 17]
```diff
@@ -89,4 +92,29 @@ def id(self):
...
+        # check author
+        if self.author and self.author in blacklists.UNTRUSTED_AUTHORS:
+            author_score = 1
+
+        # check source
+        if self.source and self.source in backlists.UNTRUSTED_SOURCES:
+            source_score = 1
+
+        # check content
+        file = open(os.path.dirname(__file__)+'/../model/mnb.pkl', 'rb')
```
Comment (JJ) [inline src/article.py:114]: Not testing for errors, and also including a literal when it should be a constant.

**Review by JJ — CHANGES_REQUESTED** [review]
Besides the issues about justification, you should test your own logic, not something contained in a pickled model. If you need a simpler model for a MVP, create that model. But it's always about testing *your* code for *your* business logic.

File: docs/dev_tools.md (hunk 8 above)
Comment (arguellesm) [reply docs/dev_tools.md]: I've changed the structure of `dev_tools.md` in https://github.com/arguellesm/faux/pull/14/commits/7fa7ae2b94212e4514d805a78c706c74b4463edd to always list criteria first, then the alternatives I found and finally the option I ended up choosing.

The ones from the last objective are obviously done retroactively, but I'll keep it in mind in the future.

File: docs/dev_tools.md (hunk 9 above)
Comment (arguellesm) [reply docs/dev_tools.md]: fixed 👍🏼

File: docs/dev_tools.md (hunk 10 above)
Comment (arguellesm) [reply docs/dev_tools.md]: I removed this criterion from the list 👍🏼

File: docs/dev_tools.md (hunk 11 above)
Comment (arguellesm) [reply docs/dev_tools.md]: When listing criteria I don't expect the final tool to meet all of them. They're meant as a guide to give some alternatives preference over others.

File: pyproject.toml (hunk 15 above)
Comment (arguellesm) [reply pyproject.toml:13]: I believe they're regular dependencies as they are always needed for the product to function properly.

File: src/article.py (hunk 16 above)
Comment (arguellesm) [reply src/article.py]: I did the model myself as described in [model.ipynb](https://github.com/arguellesm/faux/blob/7fa7ae2b94212e4514d805a78c706c74b4463edd/model/model.ipynb) from  https://github.com/arguellesm/faux/pull/14/commits/131f57aac56d2e1b1bfbdbc55d39f625370d9df6. At the end of that notebook, the model is pickled to use it in the project.

File: src/article.py (hunk 17 above)
Comment (arguellesm) [reply src/article.py]: changed 👍🏼

**Comment by arguellesm** [comment]
@JJ ready for review

File: docs/dev_tools.md [hunk 18]
```diff
@@ -1,44 +1,128 @@
 ## Development tools and why
 
-### [Poetry](https://python-poetry.org/)
+### Dependency management
```
Comment (JJ) [inline docs/dev_tools.md:3]: No one is requesting you to select a tool for dependency management. Most languages have just one, so it's not really worth the while.

File: docs/dev_tools.md [hunk 19]
```diff
@@ -1,44 +1,128 @@
...
+one. 
+
+
+### Test framework
+
+A test framework will allow us to structure tests and execute 
+them comfortably.
+
+These were the considered criteria:
+- Common within the Python community
```
Comment (JJ) [inline docs/dev_tools.md:76]: None of these criteria are really measurable or have any relation to your own project.

File: docs/dev_tools.md [hunk 20]
```diff
@@ -1,44 +1,128 @@
...
+  simple way of writing tests.
+  
+- [**Unittest**](https://docs.python.org/3/library/unittest.html)
+
+  Unittest is a built-in module in Python that comes with many
+  assertion methods. It's a bit less intuitive than Pytest as
+  tests need to be inside a class that inherits from 
+  `unittest.TestCase`.
+
+Even though Unittest is already included in Python, we opted
```
Comment (JJ) [inline docs/dev_tools.md:98]: Not comparable, not by a long shot. pytest is a test runner with a few assertions (mainly to complement assert), unittest is a testing framework without a test runner. Also, there are other tests runners and assertion libraries.

File: pyproject.toml (hunk 14 above)
Comment (JJ) [reply pyproject.toml:12]: Not clear why you need to pickle the model.

File: pyproject.toml (hunk 15 above)
Comment (JJ) [reply pyproject.toml:13]: No, they're not. Testing is not a production operation, it's a dev operation.

File: src/article.py [hunk 21]
```diff
@@ -1,5 +1,35 @@
...
+    """
+    pass
+
+class ArticleDependencyError(ModuleNotFoundError):
+    """
+    Missing dependencies error.
+    """
+    pass
+
+class CantLoadContentModelError(FileNotFoundError):
```
Comment (JJ) [inline src/article.py:20]: Not developing or testing this. It's essential that you test exceptions.

**Review by JJ — CHANGES_REQUESTED** [review]
Still some problems.
1. Try to establish some real criteria for choosing libraries. We talked about that yesterday in class.
2. Try to open up a bit your horizon of possible tools.
3. If you don't really test your own business logic, but an external model, you should at least develop exceptions and test them.
4. The "predict" function does several different things. It should have separated functions (separation of concerns) that should be tested independently.
5. Please try and understand the difference between production and development dependencies and use them accordingly.

File: docs/dev_tools.md (hunk 19 above)
Comment (arguellesm) [reply docs/dev_tools.md:76]: I changed this and tried to add ways of measuring them. They still don't have any particular relation to my project, but that's because I don't think the project needs anything too specific from the test runner or assertion library.

File: src/article.py (hunk 21 above)
Comment (arguellesm) [reply src/article.py:20]: Done.

**Comment by arguellesm** [comment]
Just a few comments. I believe I improved the biggest issues pointed out before:

1. Criteria are better know, with objective and measurable attributes.
2. I broaden my range of possible tools searching for different assertion libraries that I didn't consider before.
3. Exceptions are being tested now. There is one exception (no pun intended) to this: `ArticleDependencyError` will raise when a dependency is missing. This can't happen with the current Poetry setup, which is also used to run tests.
4. The `predict` function is now split up into three different ones, that are also tested on their own: `rate_author`, `rate_source` and `rate_content`.
5. Dev dependencies now include Invoke, Pytest and Pytruth.

**Comment by JJ** [comment]
⛹ Revisores → @Mapachana @Paszser @Parka015

File: data/generate-df.py (hunk 1 above)
Comment (JJ) [reply data/generate-df.py]: Best practices should be applied all across the project, not only in modules.

File: docs/dev_tools.md (hunk 11 above)
Comment (JJ) [reply docs/dev_tools.md]: There's no test runner included in Python. No use to put up criteria that's going to be met by no tool.

File: src/article.py (hunk 16 above)
Comment (JJ) [reply src/article.py]: You need to do it *as part of the project*, with business logic that's part of the current minimally viable product. That issue is related to #12, which is not related to any HU whatsoever. If it's not advancing an US, it's not adding value to the customer.
Remember, this objective is not about testing. It's about advancing your project to a minimally viable product that includes business logic and can be, from now on, tested with the also included infrastructure.

File: docs/dev_tools.md (hunk 19 above)
Comment (JJ) [reply docs/dev_tools.md:76]: Projects *always* need something specific, if only because they have been created at a specific point in time and should use the best practices available at that specific time.

File: docs/dev_tools.md (hunk 4 above)
Comment (JJ) [reply docs/dev_tools.md]: > Just a few comments. I believe I improved the biggest issues pointed out before:
> 
>     1. Criteria are better know, with objective and measurable attributes.
> 
>     2. I broaden my range of possible tools searching for different assertion libraries that I didn't consider before.
> 
>     3. Exceptions are being tested now. There is one exception (no pun intended) to this: `ArticleDependencyError` will raise when a dependency is missing. This can't happen with the current Poetry setup, which is also used to run tests.
> 
>     4. The `predict` function is now split up into three different ones, that are also tested on their own: `rate_author`, `rate_source` and `rate_content`.
> 
>     5. Dev dependencies now include Invoke, Pytest and Pytruth.

Not a single one of the new commits refer to any issue. If they are addressing the PR comments, they should refer to them.

File: data/README.md [hunk 22]
```diff
@@ -0,0 +1,26 @@
...
+health fake news with domain specific pre-trained 
+model](https://www.sciencedirect.com/science/article/pii/S2666285X21000662).
+
+You can download the original dataset [here](https://github.com/EnyanDai/FakeHealth).
+
+
+### Processing 
+
+The original dataset is not in CSV format and has many fields we are not 
+currently intereted in. We've extracted the following fields of interest and
```
Comment (JJ) [inline data/README.md:21]: typo

File: data/README.md [hunk 23]
```diff
@@ -0,0 +1,26 @@
...
+turned it into a single `.csv` file (which is not included in this repository).
+
```
Comment (JJ) [inline data/README.md:23]: At this point, you should have probably just modeled what's in that data and/or faked it. If you decide to process it, you should have created an issue for it, included in a milestone, and associated it to an US

File: data/generate-df.py [hunk 24]
```diff
@@ -0,0 +1,104 @@
+import pandas as pd
```
Comment (JJ) [inline data/generate-df.py:1]: Not a single line of this file is associated with an issue. Working with issues was essential since objective 2. Again, you don't need real data to test, test data can simply be fake data. If it's the structure you're looking for, you should have created an issue to model that.

File: data/generate-df.py [hunk 25]
```diff
@@ -0,0 +1,104 @@
...
+	"""
+
+	with open(path) as f:
+		file = json.load(f)
+
+	if 'author' in file.keys():
+		author = file['author']
+
+	if 'authors' in file.keys():
+		author = file['authors']
```
Comment (JJ) [inline data/generate-df.py:29]: This is the consequence of not creating issues for modeling the problem domain. You're putting author and authors in the same output variable, when quite obviously one of them is singular and the other plural. You're also extracting information without any reference of what or how it's going to be used in your business logic. At this point in the project, it's essential that you understand, model, solve what's going or not going to be included as classes in your project.

File: data/generate-df.py [hunk 26]
```diff
@@ -0,0 +1,104 @@
...
+	return author
+
+
+def generate_df():
+	"""
+	Generates .csv dataframe.
+	"""
+
+	# json fields
+	headline 	= []
```
Comment (JJ) [inline data/generate-df.py:40]: All these variables are in singular, while they are initialized with an array. Clean code calls for "semantic" identifiers, that really point to what they hold or contain.

File: data/generate-df.py [hunk 27]
```diff
@@ -0,0 +1,104 @@
...
+
+	# get history articles information
+	for i in history:
+		path = 'dataset/content/HealthStory/' + i['news_id'] + '.json'
+
+		if not os.path.exists(path):
+			continue
+
+		text.append(get_text(path))
+		author.append(get_author(path))
```
Comment (JJ) [inline data/generate-df.py:79]: [`append`](https://towardsdatascience.com/append-in-python-41c37453400) appends a single item; if it's the result of the original "authors" field you either are going to add an array as a single item, or are going to add a (possibly comma) separated list of authors as a single one. Not clear what's the case here, but again this is a reason why you need to *always* add to the repo according to HUs/issues.

File: docs/dev_tools.md (hunk 18 above)
Comment (JJ) [reply docs/dev_tools.md:3]: Again, not addressed. You should have merged to main all this part and eliminated it from here.

File: docs/dev_tools.md [hunk 28]
```diff
@@ -1,44 +1,190 @@
...
+  simple.
+
+
+_doit_ and Invoke seemed very similar to one another. When 
+looking at examples from their respective manuals, we found
+**Invoke** slightly simpler and cleaner, thus deciding on that
+one. 
+
+
+### Test framework
```
Comment (JJ) [inline docs/dev_tools.md:70]: We ask for two things (please check out the revised objective description):
* a *test runner*
* an *assertion library*

Talking about testing frameworks is misleading, because it can include everything, from mocking  libraries to runner-assertion-coverage combos.

File: docs/dev_tools.md [hunk 29]
```diff
@@ -1,44 +1,190 @@
...
+These were the considered criteria:
+- Common within the Python community
+- Included in Python
+- Ease of use
+- Simplicity
+
+When searching for test frameworks, we searched for `python + 
+[term]` and `python + [term] + stackoverflow`. These were the 
+main results:
+
```
Comment (JJ) [inline docs/dev_tools.md:84]: Again, you shouldn't. You should have tried and understand first, as we told in class, what's a test runner, and what are assertion libraries.

File: docs/dev_tools.md (hunk 20 above)
Comment (JJ) [reply docs/dev_tools.md:98]: Not addressed.

File: src/article.py [hunk 30]
```diff
@@ -1,5 +1,35 @@
-import globals
+from src import globals, blacklists
```
Comment (JJ) [inline src/article.py:1]: Best practices dictate that the module should be in a directory with the name of the project. It should never be "src"

File: src/article.py [hunk 31]
```diff
@@ -32,61 +62,99 @@ def __init__(self, headline='', author='', content='', topic='', source=''):
...
+            raise ArticleValueError('An article must include a headline or content')
 
         if(len(headline) < globals.HEADLINE_MAX_SIZE):
-            self._headline = headline
+            self.headline = headline
         else:
-            raise ValueError('Headline cannot be larger than ' + str(globals.HEADLINE_MAX_SIZE))
+            raise ArticleValueError('Headline cannot be larger than {}'.format(globals.HEADLINE_MAX_SIZE))
         
         if(globals.AUTHOR_MAX_LENGHT > len(author)):
```
Comment (JJ) [inline src/article.py:72]: typo. Also, don't understand why tests are not all here.

File: test/article_test.py (hunk 7 above)
Comment (JJ) [reply test/article_test.py:12]: Still don't belong to that. if you use truth, you should also use it here.

File: tasks.py [hunk 32]
```diff
@@ -16,3 +16,12 @@ def check(c, format='colorized'):
...
     c.run('pylint -E -f {} {}'.format(format, files))
+
+
+@task
+def test(c):
+    """
+    Run Pytest tests.
+    """
+
+    c.run('python -m pytest test')
```
Comment (JJ) [inline tasks.py:27]: Why not simply pytest?

**Review by JJ — APPROVED** [review]
I wouldn't probably pass this if you had the chance to send it again. But you don't, so it's either hit or miss. Main positive point is to have tried and be original with the assertion library, and try to test the business logic (although, again, you need to use only your own code). There are many negative points that have been pointed out, and that will probably jeopardize the fulfillment of the rest of the objectives if you don't address them.

File: data/README.md (hunk 23 above)
Comment (arguellesm) [reply data/README.md:23]: Addressed below

File: data/generate-df.py (hunk 24 above)
Comment (arguellesm) [reply data/generate-df.py:2]: Addressed below

File: data/generate-df.py (hunk 25 above)
Comment (arguellesm) [reply data/generate-df.py]: Addressed below

File: data/generate-df.py (hunk 26 above)
Comment (arguellesm) [reply data/generate-df.py]: Addressed below

File: data/generate-df.py (hunk 27 above)
Comment (arguellesm) [reply data/generate-df.py]: It adds an array of authors. I've changed the variable name to `authors` to make it more descriptive.

File: docs/dev_tools.md (hunk 28 above)
Comment (arguellesm) [reply docs/dev_tools.md]: Changed 👍🏼

File: src/article.py (hunk 30 above)
Comment (arguellesm) [reply src/article.py]: Changed name to `faux` 👍🏼

File: src/article.py (hunk 31 above)
Comment (arguellesm) [reply src/article.py]: All tests are in the `/test` directory. I didn't fix the typo because I couldn't find it.

File: tasks.py (hunk 32 above)
Comment (arguellesm) [reply tasks.py]: I left it like that because it added the current path to `sys.path`, but I don't really use that. Changed to plain `pytest` 👍🏼

**Comment by arguellesm** [comment]
Addressing the issues in `generate_df.py`:

- This file shouldn't have been included here in the first place. It's a script to (sort of) turn a JSON file into a CSV dataframe that 
it's not a requirement for the project. However, I opted for including it for documentation and replication purposes, so that anyone could see where the `.csv` file came from. I now see several problems with this:
  - There is no need to have a 'proper dataset' at this stage and   I should have used that data as dummy or fake data as you suggest.
  - By doing so, there will be no point in documenting that, not to mention providing the means to replicate it.
  - The script was thought of as a tool to generate something else and be discarded later, not as an integral part of the project, and for that reason it shouldn't have been added to the repository.
  
- On top of that, there are many issues with the `generate_df.py` script and the original JSON dataset. It was difficult to extract meaningful data because some categories would have multiple names (like 'author' and 'authors'), some keys wouldn't be present for all instances, and many values were blank. This and the fact that the script wasn't initially meant to be shared led to a messy and not very thoughtful script. It was also a mistake to rush and force having "real" data, as this could have been easily avoided with fake data, as commented before.

- I've modified a few things from `generate_df.py` to answer some of your comments, although I believe the best option would be to delete the file and consider the model was trained on fake data. Even further, there should be no data to train the model on as the model should be replaced by another implementation of the business logic that didn't require external libraries.

**tl;dr** I've made some changes to `generate_df.py` to address your comments. Regardless, I believe the best alternative would be to consider the model was based on fake/dummy data and delete `generate_df.py`.
